frame/bounties/src/lib.rs

```diff
@@ -653,12 +780,596 @@ decl_module! {
+		/// we will slash the curator when possible.
+		///
+		/// If the origin is the subcurator, we take this as a sign they are
+		/// unable to do their job and they willingly give up.
+		/// We could slash them, but for now we allow them to recover their
+		/// deposit and exit without issue. (We may want to change this
+		/// if it is abused.)
+		///
+		/// Finally, the origin can be anyone if and only if the subcurator
+		/// is "inactive". This allows anyone in the community to call out
+		/// that a subcurator is not doing their due diligence, and
+		/// we should pick a new subcurator. In this case the subcurator
+		/// should also be slashed.
+		///
+		/// State of subbounty is moved to Funded state
+		/// on successful call completion.
+		///
+		/// - `bounty_id`: ID pair Bounty ID.
+		/// - `subbounty_id`: ID pair SubBounty ID to cancel.
+		#[weight = 10_000]
+		fn unassign_subcurator(
+			origin,
+			#[compact] bounty_id: BountyIndex,
+			#[compact] subbounty_id: BountyIndex,
+		) {
+			let maybe_sender = ensure_signed(origin.clone())
+				.map(Some)
+				.or_else(|_| T::RejectOrigin::ensure_origin(origin).map(|_| None))?;
+
+			// Ensure parent bounty is Active & get status of curator
+			let ( master_curator, update_due) = Self::ensure_bounty_active(bounty_id)?;
+
+			// Ensure subbounty is in expected state
+			SubBounties::<T>::try_mutate_exists(bounty_id,
+				subbounty_id,
+				|maybe_subbounty| -> DispatchResult {
+
+				let mut subbounty = maybe_subbounty
+					.as_mut()
+					.ok_or(Error::<T>::InvalidIndex)?;
+
+				let slash_curator = |arg_curator: &T::AccountId, curator_deposit: &mut BalanceOf<T>| {
+					let imbalance = T::Currency::slash_reserved(arg_curator, *curator_deposit).0;
+					T::OnSlash::on_unbalanced(imbalance);
+					*curator_deposit = Zero::zero();
+				};
+
+				match subbounty.status {
+					SubBountyStatus::Added => {
+						// No curator to unassign at this point.
+						return Err(Error::<T>::UnexpectedStatus.into())
+					}
+					SubBountyStatus::SubCuratorProposed { ref subcurator } => {
+						// A curator has been proposed, but not accepted yet.
+						// Either `RejectOrigin`, curator or the proposed subcurator
+						// can unassign the subcurator.
+						ensure!(maybe_sender
+							.map_or(true, |sender| sender == *subcurator || sender == master_curator),
+							BadOrigin
+						);
+					},
+					SubBountyStatus::Active { ref subcurator } => {
+						// The bounty is active.
+						match maybe_sender {
+							// If the `RejectOrigin` is calling this function, slash the curator.
+							None => {
+								slash_curator(subcurator, &mut subbounty.curator_deposit);
+								// Continue to change bounty status below...
+							},
+							Some(sender) => {
+								// If the sender is not master curator or the subcurator,
+								// wait for update_due block number & slash the subcurator.
+								if sender != *subcurator && sender != master_curator {
```

Comment:
I think the logic is correct but this is a bit confusing.

you can make it:

```
if sender == subcurator {
  // subcurator give up role
} else if sender == master_curator {
  // slash
} else {
  // check for expiry 
}
```

Reply:
Sorry for the Messiness, Thanks for pointing it out :-), removed the clutters.
I hope now its clean.